### PR TITLE
DAOS-6501 dfs: Add a dfs_open_stat() function which fills in initial stat() info.

### PR DESCRIPTION
**Pull request by ashleypittman**
Add a dfs_open_stat() private call to pass out metadata for newly created files.  This removes the need to call dfs_ostat() after creating files from dfuse, so leads to a big performance benefit.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

**Comment by ashleypittman** [comment]
Not really for landing this one, but I think it's worth benchmarking to see if we can do something.

File: src/client/dfs/dfs_internal.h [hunk 1]
```diff
@@ -59,6 +59,12 @@ dfs_get_sb_layout(daos_key_t *dkey, daos_iod_t *iods[], int *akey_count,
 void
 dfs_free_sb_layout(daos_iod_t *iods[]);
 
+/* as dfs_open() but takes a stbuf */
+int
+dfs_open2(dfs_t *dfs, dfs_obj_t *parent, const char *name, mode_t mode,
```
Comment (mchaarawi) [inline src/client/dfs/dfs_internal.h:64]: let's call this dfs_open_stat() ?

File: src/client/dfs/dfs_internal.h [hunk 2]
```diff
@@ -59,6 +59,12 @@ dfs_get_sb_layout(daos_key_t *dkey, daos_iod_t *iods[], int *akey_count,
...
+	  int flags, daos_oclass_id_t cid, daos_size_t chunk_size,
+	  const char *value, struct stat *stbuf, dfs_obj_t **obj);
```
Comment (mchaarawi) [inline src/client/dfs/dfs_internal.h:66]: would prefer if stbuf was last argument

File: src/client/dfs/dfs_internal.h (hunk 1 above)
Comment (ashleypittman) [reply src/client/dfs/dfs_internal.h]: Done.

File: src/client/dfs/dfs_internal.h (hunk 2 above)
Comment (ashleypittman) [reply src/client/dfs/dfs_internal.h]: Done.  obj is also an out arg so I'd left that one last.

**Review by mchaarawi — APPROVED** [review]

File: src/client/dfs/dfs.c [hunk 3]
```diff
@@ -2326,6 +2349,9 @@ dfs_open(dfs_t *dfs, dfs_obj_t *parent, const char *name, mode_t mode,
 	else if (!S_ISDIR(parent->mode))
 		return ENOTDIR;
 
+	if (stbuf && !(flags & O_CREAT))
+		return ENOTSUP;
+
```
[the remark this answers is not in the document]
Comment (daltonbohning) [reply src/client/dfs/dfs.c:2354]: Update: Actually, I see we get the stat info from `entry` below here. So is there some other reason why we can't support this?

File: src/client/dfs/dfs.c [hunk 4]
```diff
@@ -748,8 +749,8 @@ open_file(dfs_t *dfs, daos_handle_t th, dfs_obj_t *parent, int flags,
 		return ENOENT;
 
 fopen:
-	if (!S_ISREG(entry.mode)) {
-		D_FREE(entry.value);
+	if (!S_ISREG(entry->mode)) {
+		D_FREE(entry->value);
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/dfs/dfs.c:753]: entry->value is the contents of any symlink, so we could check for S_ISLNK before the free but it wouldn't help, we typically don't check for things like this as the cost of D_FREE(NULL) is very low, and it provides useful debugging into flow though the code, as well as introduces scope for bugs and increasing code complexity.

Essentially this path is "the user tried to open a file and something else is there" so we just cleanup and return.

File: src/client/dfs/dfs.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/dfs/dfs.c:2354]: We possibly could for open_file, I'm working on adding test coverage for this currently, particularly where O_CREAT is set but not O_EXCL and the file already exists.

open_dir() works with O_CREAT as expected, but the rest of open_dir() is only called for / by dfs_mount() so could probably just be a seperate function.

open_symlink() doesn't have any code to support the no O_CREAT path at all.

At the minute this is an internal function, and only called from dfuse, and always with O_CREAT, but it's also an odd function in that it's open, but specifies *what* it's trying to open, it should probably be dfs_create() rather than dfs_open() or dfs_open_stat() given what it does, but this is still an active area of development, and why it's best not to commit to interfaces too early in the development cycle IMO.

In short, it's not that we can't support this, it's that we don't need to for current use-cases and are still working through getting the API correct.

File: src/client/dfs/dfs.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/dfs/dfs.c:753]: Then again, fetch_entry is called with false to fetch_symlink so this should always be NULL so I could probably change this to assert.

[the remark this answers is not in the document]
Comment (daltonbohning) [reply src/client/dfs/dfs.c:753]: Ah, I didn't realize `D_FREE()` handled NULL. I was thinking that `free(NULL)` would be invalid.
At one point, I do remember that some `asserts` were removed in favor of `D_FREE`, though I can't remember when and where... 
Anyway, if the `D_FREE` here isn't an issue, no worries.

File: src/client/dfs/dfs.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (daltonbohning) [reply src/client/dfs/dfs.c:2354]: I guess I don't object to that 👍 
Really I was thinking that if we _**did**_ have the stat info available for all cases, then we could reduce complexity here by removing this check, and setting stat info if and only if `stbuf` was passed in - similar to dfs_lookup, etc.
But you're right, this is just internal, so we can do whatever we want with it.

**Review by daltonbohning — APPROVED** [review]

File: src/client/dfs/dfs.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/dfs/dfs.c:753]: free() and D_FREE() have always handled NULL, and in fact we encourage D_FREE() to be used without a NULL check.

File: src/client/dfs/dfs.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/dfs/dfs.c:2354]: If we did have it, then we should use it and not discard it, but we neither have it in this case, or have a code-path that could get to this case.

File: src/client/dfs/dfs.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (daltonbohning) [reply src/client/dfs/dfs.c:753]: You're right. Semantically, I wrote that wrong. I was thinking of this:
```
char *s;
free(s); // s is not initialized to NULL
```
But we _**DO**_ initialize this to NULL.
I think this line of code is fine as-is.

[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/dfs/dfs.c:753]: Both the compiler and valgrind checks (on the client) would catch that thankfully.
